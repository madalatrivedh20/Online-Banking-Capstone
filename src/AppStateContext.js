import { createContext, useState, useContext } from "react";

const AppStateContext = createContext();

export const AppStateProvider = ({ children }) => {
  const [isAuthenticated, setIsAuthenticated] = useState(false);
  const [issocialAuthenticated, setIssocialAuthenticated] = useState(false);
  const [user, setUser] = useState({});
  const [allUsers, setAllUsers] = useState([]);

  const value = {
    isAuthenticated,
    setIsAuthenticated,
    user,
    issocialAuthenticated,
    setIssocialAuthenticated,
    setUser,
    allUsers,
    setAllUsers
  };
  return <AppStateContext.Provider value={value}>{children}</AppStateContext.Provider>;
};

const useAppState = () => {
  const context = useContext(AppStateContext);

  if (context === undefined) {
    throw new Error("useAppState must be used within AppStateContext");
  }

  return context;
};

export default useAppState;
