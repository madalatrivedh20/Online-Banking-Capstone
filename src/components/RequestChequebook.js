import '../style/Login.css';
import useAppState from '../AppStateContext';
import { useState } from 'react';
import validator from 'validator';
import { toast } from 'react-toastify';
import { useNavigate } from 'react-router-dom';
import { v4 as uuid } from 'uuid';


// Component that handles raising a request for the new cheque book

const RequestChequebook = () => {
  const navigate = useNavigate();
  const FormHeader = props => (
    <div>
      <h2 id="headerTitle">{props.title}</h2>
    </div>
  );
  const { isAuthenticated, user } = useAppState();

// Function invoked on submission of the form with the details like address

  const submitHandler = (e) => {
    e.preventDefault();
    if (!validator.isAlpha(data['state'][0])) return toast.error("Invalid State");
    if (!validator.isAlpha(data['city'][0])) return toast.error("Invalid City");
    let user_id = user.id;
    let unique_id = uuid();
    let cheque_id = unique_id.slice(0, 8);
    const current = new Date();
    const date = `${current.getDate()}/${current.getMonth() + 1}/${current.getFullYear()}`;
    let issue_date = `${current.getDate() + 3}/${current.getMonth() + 1}/${current.getFullYear()}`;

    const requestOptions = {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        'firstname': user.firstname, 'acctype': user.acctype,
        'accno': user.accno, 'email': user.email, 'password': user.password, 'sectype': user.sectype, 'secans': user.secans, 'transactions': [], 'balance': '30000$'
        , "cheque": {
          'accno': data['accno'][0], 'state': data['state'][0],
          'city': data['city'][0], 'zipcode': data['zipcode'][0], 'address': data['address'][0], 'priority': data['priority'][0],
          'chequebook_id': cheque_id, 'requested_date': date, 'issued_date': issue_date
        }
      })
    };
    fetch(`http://localhost:3000/users/${user_id}`, requestOptions)
      .then(response => response.json())
      .then(data => console.log(data));

    toast.success("Checkbook request raised Successfully");
    setData({
      accno: user.accno,
      state: "",
      city: "",
      zipcode: "",
      address: "",
      priority: "",
      checkbook_id: "",
      requested_date: "",
      issue_date: ""
    });

    navigate('/balance');



  };
  const [data, setData] = useState({
    accno: user.accno,
    state: "",
    city: "",
    zipcode: "",
    address: "",
    priority: ""
  });
  const changeHandler = e => {
    setData({ ...data, [e.target.name]: [e.target.value] });

  };


  return (
    <div className="forms">
      <FormHeader title="Request a New Checkbook" />
      <div>
        <form onSubmit={submitHandler}>
          <div className="row">
            <label>Account Number</label>
            <input disbaled required type="text" name="accno" placeholder="Account Number" value={user.accno} onChange={changeHandler} />
          </div>
          <FormHeader title="Billing Address" />
          <div className="row">
            <label>State</label>
            <input required type="text" name="state" placeholder="Enter your State" value={data.state} onChange={changeHandler} />
          </div>

          <div className="row">
            <label>City</label>
            <input required type="text" name="city" placeholder="Enter your City" value={data.city} onChange={changeHandler} />
          </div>

          <div className="row">
            <label>Zip code</label>
            <input required type="text" name="zipcode" placeholder="Enter your Pincode" value={data.zipcode} onChange={changeHandler} />
          </div>

          <div className="row">
            <label>Address</label>
            <input required type="textarea" name="address" placeholder="Enter your Address" value={data.address} onChange={changeHandler} />
          </div>
          <div className="row">
            <label>Priority</label>

            <select name="priority" required value={data.priority} onChange={changeHandler}>
              <option value={"Select your Priority"}>Select your Priority</option>
              <option value={"Urgent"}>Urgent (1-2 Days)</option>
              <option value={"High"}>High (3-5 Days)</option>
              <option value={"Medium"}>Medium (5-7 Days)</option>
            </select>
          </div>
          <div id="button" className="row">
            <button>Raise a request</button>

          </div>
        </form>

      </div>


    </div>

  );
};

export default RequestChequebook;