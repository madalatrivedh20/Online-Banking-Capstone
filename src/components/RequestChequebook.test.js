import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import renderer from 'react-test-renderer';
import RequestChequebook from './RequestChequebook';

import {
  BrowserRouter as Router,
  Routes,
  Route
} from "react-router-dom";
import { AppStateProvider } from '../AppStateContext';

it('Snapshot Test', () => {
  const component = renderer.create(
    <AppStateProvider>
      <Router>
        <RequestChequebook />
      </Router>
    </AppStateProvider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders the page', async () => {
  render(
    <AppStateProvider>
      <Router>
        <RequestChequebook />
      </Router>
    </AppStateProvider>
  );

  expect(screen.getByText("Request a New Checkbook")).toBeInTheDocument();
  expect(screen.getByText("Billing Address")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Account Number")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Enter your State")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Enter your City")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Enter your Pincode")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Enter your Address")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Select Priority")).toBeInTheDocument();
  expect(screen.getByRole("button", { name: "Raise a request" })).toBeInTheDocument();
});