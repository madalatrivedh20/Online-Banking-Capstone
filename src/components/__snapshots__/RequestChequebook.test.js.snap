// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
<div
  className="forms"
>
  <div>
    <h2
      id="headerTitle"
    >
      Request a New Checkbook
    </h2>
  </div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <label>
          Account Number
        </label>
        <input
          disbaled={true}
          name="accno"
          onChange={[Function]}
          placeholder="Account Number"
          required={true}
          type="text"
        />
      </div>
      <div>
        <h2
          id="headerTitle"
        >
          Billing Address
        </h2>
      </div>
      <div
        className="row"
      >
        <label>
          State
        </label>
        <input
          name="state"
          onChange={[Function]}
          placeholder="Enter your State"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <label>
          City
        </label>
        <input
          name="city"
          onChange={[Function]}
          placeholder="Enter your City"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <label>
          Zip code
        </label>
        <input
          name="zipcode"
          onChange={[Function]}
          placeholder="Enter your Pincode"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <label>
          Address
        </label>
        <input
          name="address"
          onChange={[Function]}
          placeholder="Enter your Address"
          required={true}
          type="textarea"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <label>
          Priority
        </label>
        <select
          name="priority"
          onChange={[Function]}
          placeholder="Select Priority"
          required={true}
          value=""
        >
          <option
            value="Select your Priority"
          >
            Select your Priority
          </option>
          <option
            value="Urgent"
          >
            Urgent (1-2 Days)
          </option>
          <option
            value="High"
          >
            High (3-5 Days)
          </option>
          <option
            value="Medium"
          >
            Medium (5-7 Days)
          </option>
        </select>
      </div>
      <div
        className="row"
        id="button"
      >
        <button>
          Raise a request
        </button>
      </div>
    </form>
  </div>
</div>
`;
